crypto/init.c

```diff
@@ -729,3 +683,52 @@ int OPENSSL_atexit(void (*handler)(void))
     return 1;
 }
 
+/*
+ * Library  destructor (a.k.a. _fini()). Code here is derived
+ * from stuff found in providers/fips/self_test.c
+ *
+ * We do kind of  best effort cleanup. If given platform/compiler
+ * does not support module destructors then its up to application
+ * to explicitly call OPENSSL_cleanup();
+ */
+#if defined(__GNUC__) && !defined(_AIX)
+
+void ossl_call_cleanup(void);
+
+void __attribute__((destructor))
+ossl_call_cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+
+#elif (defined(__sun) || defined(_AIX)) && !defined(__GNUC__)
+void ossl_call_cleanup(void);
+# pragma fini(ossl_call_cleanup)
+void ossl_call_cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+
+#elif defined(__hpux)
+# pragma fini "ossl_call_cleanup"
+void ossl_call_cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+#elif defined(__TANDEM)
+/* Method automatically called by the NonStop OS prior to unloading the DLL */
+void __TERM__cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+#else
+void _fini(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+#endif
```

Comment:
I prefer the providers/fips/self_test.c way of doing this which avoids the duplication of the same function definition over and over again. The way we have it here we have 5 variants of essentially the same function implementation. Any maintainer changing this code will have to make sure it is reflected 5 times - which seems like a recipe for things getting out of sync with each other over time.

Reply:
I see your point here. The reason I diverged from `providers/fips/self/self_test.c` was I got lost in a maze of defs/ifdefs/undefs. Just say yes to: make code in this PR consistent with `providers/fips/self/self_test.c`  and I'll do it.

Reply:
I would prefer it that way, but I don't insist on it. Perhaps other reviewers might have an opinion.

Reply:
I am OK with it as is although I am missing something for Windows here.